apps/alerts/api.py

```diff
@@ -0,0 +1,160 @@
+            "secret": getenv("AUTH_CLIENT_SECRET"),
+            "email": email,
+            "course": course,
+        },
+    ).json()
+
+
+def clear_collection(db: "firestore.Client", ref):
+    batch = db.batch()
+    cnt = 0
+    for document in ref.stream():
+        batch.delete(document.reference)
+        cnt += 1
+        if cnt > 400:
+            batch.commit()
+            batch = db.batch()
+            cnt = 0
+    batch.commit()
+
+
+def process_ok_exam_upload(db: "firestore.Client", data, secret):
+    """
+    data: {
+        "exam_name": string,
+        "students": [
+            {
+                "email": string,
+                "questions": [
+                    {
+                        "student_question_name": string,
+                        "canonical_question_name": string,
+                        "student_question_text": string,
+                        "start_time": int,
+                        "end_time": int,
+                    }
+                ],
+                "start_time": int,
+                "end_time": int,
+            }
+        ]
+        "questions": [
+            {
+                "question_name": string,
+                "canonical_question_text": string,
+            }
+        ],
+    }
+    """
+    course = data["exam_name"].split("-")[0]
+    email = get_email_from_secret(secret)
+    if not is_admin(email, course):
+        abort(403)
+    db.collection("exam-alerts").document(data["exam_name"]).set(
+        {"questions": data["questions"]}
+    )
+    ref = (
+        db.collection("exam-alerts").document(data["exam_name"]).collection("students")
+    )
+    clear_collection(db, ref)
+
+    batch = db.batch()
+    cnt = 0
+    for student in data["students"]:
+        doc_ref = ref.document(student["email"])
+        batch.set(doc_ref, student)
+        cnt += 1
+        if cnt > 400:
+            batch.commit()
+            batch = db.batch()
+            cnt = 0
+    batch.commit()
+
+    ref = db.collection("exam-alerts").document("all")
+    data = ref.get().to_dict()
+    if data["exam_name"] not in data["exam-list"]:
+        data["exam-list"].append(data["exam_name"])
+    ref.set(data)
+
+
+def get_announcements(student_data, announcements):
+    """
+    Announcements are of the form
+    {
+        type: "scheduled" | "immediate",
+        canonical_question_name: string,
+        offset: int,
+        base: "start" | "end",
+        message: string,
+    }
+    Immediate announcements only need a message. If no question name is provided, the announcement will be
+    made relative to the exam start/end, otherwise it will be relative to the question
+    when that question starts / ends.
+
+    Any instances of the `canonical_question_name` will be replaced with the `student_question_name` when the
+    announcement is sent to students. Similarly, any exam text substitutions will be made in the message before
+    it is sent.
```

Comment:
does this happen anywhere? I don't see it

Reply:
Whoops, forgot to do that. The text substitutions are probably gonna be for examtool only, so I'll remove this sentence for now until it's implemented